specifications/ReadersWriters/ReadersWriters.tla

```diff
@@ -0,0 +1,122 @@
+-------------------------- MODULE ReadersWriters --------------------------
+
+EXTENDS FiniteSets, Naturals, Sequences
+
+CONSTANT NumActors
+
+VARIABLES
+    readers, \* set of processes currently reading
+    writers, \* set of processes currently writing
+    waiting  \* queue of processes waiting to access the resource
+
+vars == <<readers, writers, waiting>>
+
+Actors == 1..NumActors
+
+ToSet(s) == { s[i] : i \in DOMAIN s }
+
+fst(s)  == s[1]
+nfst(s) == ~s[1]
+
+WaitingToRead  == { p[2] : p \in ToSet(SelectSeq(waiting, fst)) }
+
+WaitingToWrite == { p[2] : p \in ToSet(SelectSeq(waiting, nfst)) }
+
+---------------------------------------------------------------------------
+
+(***********)
+(* Actions *)
+(***********)
+
+TryRead(actor) ==
+    /\ actor \notin WaitingToRead
+    /\ waiting' = Append(waiting, <<TRUE, actor>>)
+    /\ UNCHANGED <<readers, writers>>
+
+TryWrite(actor) ==
+    /\ actor \notin WaitingToWrite
+    /\ waiting' = Append(waiting, <<FALSE, actor>>)
+    /\ UNCHANGED <<readers, writers>>
+
+Read(actor) ==
+    /\ readers' = readers \union {actor}
+    /\ waiting' = Tail(waiting)
+    /\ UNCHANGED writers
+
+Write(actor) ==
+    /\ readers = {}
+    /\ writers' = writers \union {actor}
+    /\ waiting' = Tail(waiting)
+    /\ UNCHANGED readers
+
+StopActivity(actor) ==
+    IF actor \in readers
+    THEN /\ readers' = readers \ {actor}
+         /\ UNCHANGED <<writers, waiting>>
+    ELSE /\ writers' = writers \ {actor}
+         /\ UNCHANGED <<readers, waiting>>
+
+ReadOrWrite ==
```

Comment:
This should be moved above the definition of StopActivity because it "belongs" to the definitions of Read and Write. Personally, I would make the definitions of the Read and Write actions self-contained, i.e. merge them with the relevant parts of ReadOrWrite. In particular, make all pre-conditions of the action explicit in the action: now, one has "writers = {}" in ReadOrWrite and then "readers = {}" in Write. But you may feel differently about this.

Reply:
I doubt my "style" is idiomatic. I like this way (splitting up the action and enabling condition) better because, in my mind, then I can more easily think about these pieces separately.

I'll defer to your judgement here. It's not a big change...

Reply:
After going to make this change, I didn't immediately see how to do it in a nice way. Since `waiting` is a queue and we need to inspect the head to know whether to trigger `Read` or `Write`, it's not clear how to make the actions self-contained without including an inspection of the head element of `waiting` in both. Maybe this is what you are suggesting, but I'd argue that that makes it harder to read. I'm curious what you would do in this case.